src/gui/qgsmaptooledit.cpp

```diff
@@ -27,6 +27,8 @@
 QgsMapToolEdit::QgsMapToolEdit( QgsMapCanvas *canvas )
   : QgsMapTool( canvas )
 {
+  connect( QgsProject::instance(), &QgsProject::layersAdded, this, &QgsMapToolEdit::connectLayers );
```

Comment:
Thinking about it again, how about using `QgsMapCanvas::layersChanged`? Using `QgsProject::instance()` is discouraged and actually the tools always work on canvas layers?

Reply:
Then things get complicated. 

In QGIS I can edit a layer even if it's not visible on the map canvas. Also, the map canvas only gives us access to layers that are currently being displayed. 

For this PR (remove rubber bands when no editable layer is left in the QGIS project), that would mean I would have to keep a list of layers that once were added to canvas (layer registry?), and once the user stops the edits on a layer, go to that "layer registry" (not to the canvas) to ask if any layer there is still being edited. 

If we don't use such "layer registry", we would miss invisible layers' `editingStopped` signals, so still rubber bands could be on the map canvas even if no layers are being edited in the project.

As you see, using `QgsProject::layersAdded` is very convenient here.

Reply:
@m-kuhn looks like this PR stalled on this question. Can you reply please?

Reply:
Thanks for the heads up @nyalldawson 

Fair point, @gacarrillor , good explanation why the map canvas layer set is not sufficient here. 

Do we have any chance for a setter for the project, so we don't have to use the global singleton? The most straightforward thing would be to add `QgsMapCanvas::mProject`.

Reply:
@m-kuhn do you mean to use a public method `project()` from `QgsMapCanvas`? If so, are you OK if we set the main canvas' project [in this block](https://github.com/qgis/QGIS/blob/master/src/app/qgisapp.cpp#L922-L943)?

Reply:
Thant would make perfect sense for me.
@nyalldawson are you supportive of this approach?

Reply:
Yes, but for consistency make sure you also set the project for additional map canvases in QgisApp::createNewMapCanvasDock()